source/common/matcher/field_matcher.h

```diff
@@ -144,7 +144,8 @@ template <class DataType>
 class SingleFieldMatcher : public FieldMatcher<DataType>, Logger::Loggable<Logger::Id::matcher> {
 public:
   SingleFieldMatcher(DataInputPtr<DataType>&& data_input, InputMatcherPtr&& input_matcher)
-      : data_input_(std::move(data_input)), input_matcher_(std::move(input_matcher)) {}
+      : data_input_(std::move(validateInputAndMatcher(data_input, input_matcher))),
+        input_matcher_(std::move(input_matcher)) {}
```

Comment:
> may just validate the input and matcher in the constructor function rather than in the initilize list to improve the readability?

sorry, i think this is comment for this.

Reply:
Yea, here we can do it on `data_input_` and `input_matcher_ `. But i didn't do this because:
1) be consistent with `initializer_list_validate style` at places above where we cannot do validation in constructor.
2) slightly better to throw exception earlier and don't initialized those class member when it is invalid case. i.e., initialization of class members will be redundant for invalid case.

wdyt?

Reply:
I think envoy have no any preference to the style. Both styles are ok.

Specific to this patch, imo, validate the input type in the ctor function will make the code easy to read/understand. This is very important for long term maintenance.

By that way, we needn't prapagate the input to the validator and return it after the validation and finally move it.

And about the overhead, I think there should no any difference.

Reply:
Here is 2 std::move on smart pointers. So, the overhead is not just moving two pointers as it involves updating the internal state to transfer ownership.

But I don't have strong preference over `initializer_list_validate` style either.  Either style works for me and